File: protocol/reference/raven_protocol/device_revocation.py
"""RavenDeviceRevocationV1 — exact wire codec + store snapshot hash (vector freeze)."""

from __future__ import annotations

import hashlib
from dataclasses import dataclass, field

from cryptography.exceptions import InvalidSignature
from cryptography.hazmat.primitives.asymmetric.ed25519 import (
    Ed25519PrivateKey,
    Ed25519PublicKey,
)

from ._canon import lp, u64
from . import address as raven_address
from .pair_init import device_certificate_hash

MAGIC = b"RVDR1\0\0\0"
VERSION = 0x01
SUITE = 0x01
SIGNING_DOMAIN = b"rvn1/device-revocation"
STORE_SNAPSHOT_DOMAIN = b"rvn1/device-revocation/store-v1"
ADDRESS_LEN = 44
PUB_LEN = 32
HASH_LEN = 32
REVOCATION_ID_LEN = 16
SIGNATURE_LEN = 64
ID_MIN = 1
ID_MAX = 64

# Fixed header before first variable lp(device_id)
# magic(8)+version(1)+suite(1)+address(44) = 54
FIXED_PREFIX_LEN = 54


@dataclass
class DeviceRevocationV1:
    identity_address: str
    device_id: bytes  # raw UTF-8 bytes
    device_ed_pub: bytes
    device_x_pub: bytes
    device_cert_hash: bytes
    issuer_device_id: bytes
    issuer_seq: int
    revocation_id: bytes
    reason_code: int
    created_at_ms: int
    signature: bytes = field(default=b"")


def _require_id(label: str, raw: bytes) -> bytes:
    if not isinstance(raw, (bytes, bytearray)):
        raise ValueError(f"{label} must be bytes")
    raw = bytes(raw)
    if not (ID_MIN <= len(raw) <= ID_MAX):
        raise ValueError(f"{label} length must be 1..64")
    return raw


def _require_address(addr: str) -> bytes:
    if not isinstance(addr, str) or len(addr) != ADDRESS_LEN:
        raise ValueError("identity_address must be exactly 44 ASCII chars")
    decoded = raven_address.decode(addr)
    if decoded is None:
        raise ValueError("identity_address is not a valid RavenAddressV1")
    if addr != addr.lower() or not addr.startswith("rvn1"):
        raise ValueError("identity_address must be canonical lowercase")
    return addr.encode("ascii")


def signing_bytes(r: DeviceRevocationV1) -> bytes:
    device_id = _require_id("device_id", r.device_id)
    issuer = _require_id("issuer_device_id", r.issuer_device_id)
    addr = _require_address(r.identity_address)
    if len(r.device_ed_pub) != PUB_LEN or len(r.device_x_pub) != PUB_LEN:
        raise ValueError("device pubs must be 32 bytes")
    if len(r.device_cert_hash) != HASH_LEN:
        raise ValueError("device_cert_hash must be 32 bytes")
    if len(r.revocation_id) != REVOCATION_ID_LEN or r.revocation_id == bytes(REVOCATION_ID_LEN):
        raise ValueError("revocation_id must be 16 nonzero bytes")
    if not (0 <= r.reason_code <= 255):
        raise ValueError("reason_code out of range")
    if r.issuer_seq < 0 or r.issuer_seq >= 2**64:
        raise ValueError("issuer_seq out of range")
    if r.created_at_ms < 0 or r.created_at_ms >= 2**64:
        raise ValueError("created_at_ms out of range")
    return (
        SIGNING_DOMAIN
        + bytes([VERSION, SUITE])
        + addr
        + lp(device_id)
        + r.device_ed_pub
        + r.device_x_pub
        + r.device_cert_hash
        + lp(issuer)
        + u64(r.issuer_seq)
        + r.revocation_id
        + bytes([r.reason_code & 0xFF])
        + u64(r.created_at_ms)
    )


def encode(r: DeviceRevocationV1) -> bytes:
    if len(r.signature) != SIGNATURE_LEN:
        raise ValueError("signature must be 64 bytes")
    sb = signing_bytes(r)
    body = MAGIC + sb[len(SIGNING_DOMAIN) :] + r.signature
    return body


def _read_lp(buf: bytes, off: int) -> tuple[bytes, int]:
    if off + 2 > len(buf):
        raise ValueError("truncated lp")
    n = int.from_bytes(buf[off : off + 2], "big")
    off += 2
    if not (ID_MIN <= n <= ID_MAX):
        raise ValueError("lp length out of 1..64")
    if off + n > len(buf):
        raise ValueError("truncated lp body")
    return buf[off : off + n], off + n


def decode(wire: bytes) -> DeviceRevocationV1:
    if not isinstance(wire, (bytes, bytearray)):
        raise ValueError("wire must be bytes")
    wire = bytes(wire)
    if len(wire) < FIXED_PREFIX_LEN + 2 + ID_MIN + PUB_LEN * 2 + HASH_LEN + 2 + ID_MIN + 8 + 16 + 1 + 8 + 64:
        raise ValueError("wire too short")
    if wire[:8] != MAGIC:
        raise ValueError("bad magic")
    if wire[8] != VERSION or wire[9] != SUITE:
        raise ValueError("bad version/suite")
    off = 10
    addr = wire[off : off + ADDRESS_LEN]
    off += ADDRESS_LEN
    try:
        identity_address = addr.decode("ascii")
    except UnicodeDecodeError as e:
        raise ValueError("identity_address not ASCII") from e
    _require_address(identity_address)
    device_id, off = _read_lp(wire, off)
    device_ed = wire[off : off + PUB_LEN]
    off += PUB_LEN
    device_x = wire[off : off + PUB_LEN]
    off += PUB_LEN
    cert_hash = wire[off : off + HASH_LEN]
    off += HASH_LEN
    issuer_id, off = _read_lp(wire, off)
    issuer_seq = int.from_bytes(wire[off : off + 8], "big")
    off += 8
    rev_id = wire[off : off + REVOCATION_ID_LEN]
    off += REVOCATION_ID_LEN
    reason = wire[off]
    off += 1
    created = int.from_bytes(wire[off : off + 8], "big")
    off += 8
    sig = wire[off : off + SIGNATURE_LEN]
    off += SIGNATURE_LEN
    if off != len(wire):
        raise ValueError("trailing bytes")
    if rev_id == bytes(REVOCATION_ID_LEN):
        raise ValueError("revocation_id all-zero")
    return DeviceRevocationV1(
        identity_address=identity_address,
        device_id=device_id,
        device_ed_pub=device_ed,
        device_x_pub=device_x,
        device_cert_hash=cert_hash,
        issuer_device_id=issuer_id,
        issuer_seq=issuer_seq,
        revocation_id=rev_id,
        reason_code=reason,
        created_at_ms=created,
        signature=sig,
    )


def sign(r: DeviceRevocationV1, identity_ed_priv: bytes) -> DeviceRevocationV1:
    priv = Ed25519PrivateKey.from_private_bytes(identity_ed_priv)
    pub = priv.public_key().public_bytes_raw()
    # Address must match signer
    if raven_address.encode(pub) != r.identity_address:
        raise ValueError("identity_address does not match signing key")
    sig = priv.sign(signing_bytes(r))
    return DeviceRevocationV1(
        identity_address=r.identity_address,
        device_id=r.device_id,
        device_ed_pub=r.device_ed_pub,
        device_x_pub=r.device_x_pub,
        device_cert_hash=r.device_cert_hash,
        issuer_device_id=r.issuer_device_id,
        issuer_seq=r.issuer_seq,
        revocation_id=r.revocation_id,
        reason_code=r.reason_code,
        created_at_ms=r.created_at_ms,
        signature=sig,
    )


def verify(r: DeviceRevocationV1, identity_ed_pub: bytes) -> bool:
    if raven_address.encode(identity_ed_pub) != r.identity_address:
        return False
    try:
        Ed25519PublicKey.from_public_bytes(identity_ed_pub).verify(
            r.signature, signing_bytes(r)
        )
        return True
    except (InvalidSignature, ValueError):
        return False


def claim_digest(wire: bytes) -> bytes:
    return hashlib.sha256(bytes(wire)).digest()


def object_digest(wire: bytes) -> bytes:
    """Under bare packaging, equals claim_digest."""
    return claim_digest(wire)


def wire_offsets(device_id: bytes, issuer_device_id: bytes) -> dict[str, int]:
    """Document exact offsets for a concrete id pair (vector metadata)."""
    device_id = _require_id("device_id", device_id)
    issuer = _require_id("issuer_device_id", issuer_device_id)
    o = 0
    offsets = {"magic": o}
    o += 8
    offsets["version"] = o
    o += 1
    offsets["suite"] = o
    o += 1
    offsets["identity_address"] = o
    o += ADDRESS_LEN
    offsets["device_id_len"] = o
    o += 2
    offsets["device_id"] = o
    o += len(device_id)
    offsets["device_ed_pub"] = o
    o += PUB_LEN
    offsets["device_x_pub"] = o
    o += PUB_LEN
    offsets["device_cert_hash"] = o
    o += HASH_LEN
    offsets["issuer_device_id_len"] = o
    o += 2
    offsets["issuer_device_id"] = o
    o += len(issuer)
    offsets["issuer_seq"] = o
    o += 8
    offsets["revocation_id"] = o
    o += REVOCATION_ID_LEN
    offsets["reason_code"] = o
    o += 1
    offsets["created_at_ms"] = o
    o += 8
    offsets["signature"] = o
    o += SIGNATURE_LEN
    offsets["total_len"] = o
    return offsets


@dataclass(frozen=True)
class StoreClaim:
    exact_record_bytes: bytes

    @property
    def claim_digest(self) -> bytes:
        return claim_digest(self.exact_record_bytes)


@dataclass(frozen=True)
class ExhaustedMarker:
    identity_address: str
    claim_digest: bytes
    exact_record_bytes: bytes


@dataclass(frozen=True)
class CorruptMarker:
    scope: str
    reason_code: int


def canonical_store_snapshot(
    generation: int,
    claims: list[StoreClaim],
    exhausted: list[ExhaustedMarker],
    corrupt: list[CorruptMarker],
) -> bytes:
    if generation < 0 or generation >= 2**64:
        raise ValueError("generation out of range")
    claims_sorted = sorted(claims, key=lambda c: c.claim_digest)
    exh_sorted = sorted(
        exhausted, key=lambda e: (e.identity_address.encode("ascii"), e.claim_digest)
    )
    cor_sorted = sorted(corrupt, key=lambda c: c.scope.encode("utf-8"))

    out = STORE_SNAPSHOT_DOMAIN + b"\x00" + u64(generation)

    out += len(claims_sorted).to_bytes(4, "big")
    for c in claims_sorted:
        d = c.claim_digest
        b = c.exact_record_bytes
        if len(d) != 32:
            raise ValueError("bad claim digest")
        out += d + len(b).to_bytes(4, "big") + b

    out += len(exh_sorted).to_bytes(4, "big")
    for e in exh_sorted:
        addr = e.identity_address.encode("ascii")
        if len(e.claim_digest) != 32:
            raise ValueError("bad exhausted digest")
        if claim_digest(e.exact_record_bytes) != e.claim_digest:
            raise ValueError("exhausted blob digest mismatch")
        out += lp(addr) + e.claim_digest + len(e.exact_record_bytes).to_bytes(4, "big")
        out += e.exact_record_bytes

    out += len(cor_sorted).to_bytes(4, "big")
    for c in cor_sorted:
        scope = c.scope.encode("utf-8")
        if len(scope) > 0xFFFF:
            raise ValueError("corrupt scope too long")
        out += lp(scope) + bytes([c.reason_code & 0xFF])

    return out


def revocation_store_hash(
    generation: int,
    claims: list[StoreClaim],
    exhausted: list[ExhaustedMarker],
    corrupt: list[CorruptMarker],
) -> bytes:
    return hashlib.sha256(
        canonical_store_snapshot(generation, claims, exhausted, corrupt)
    ).digest()


# Re-export for generators that already hash certs via pair_init
__all__ = [
    "DeviceRevocationV1",
    "MAGIC",
    "VERSION",
    "SUITE",
    "signing_bytes",
    "encode",
    "decode",
    "sign",
    "verify",
    "claim_digest",
    "object_digest",
    "wire_offsets",
    "StoreClaim",
    "ExhaustedMarker",
    "CorruptMarker",
    "canonical_store_snapshot",
    "revocation_store_hash",
    "device_certificate_hash",
]

File: node/crates/raven-core/src/device_revocation.rs
//! RavenDeviceRevocationV1 — exact wire + store snapshot hash.
//! Spec: `protocol/RAVEN_DEVICE_REVOCATION_V1.md` (vector freeze).

use crate::address::{decode_address, encode_address};
use crate::canon::{lp, u64_be};
use crate::identity::Identity;
use sha2::{Digest, Sha256};

pub const MAGIC: &[u8; 8] = b"RVDR1\0\0\0";
pub const VERSION: u8 = 0x01;
pub const SUITE: u8 = 0x01;
pub const SIGNING_DOMAIN: &[u8] = b"rvn1/device-revocation";
pub const STORE_SNAPSHOT_DOMAIN: &[u8] = b"rvn1/device-revocation/store-v1";
const ADDRESS_LEN: usize = 44;
const ID_MIN: usize = 1;
const ID_MAX: usize = 64;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct DeviceRevocationV1 {
    pub identity_address: String,
    pub device_id: Vec<u8>,
    pub device_ed_pub: [u8; 32],
    pub device_x_pub: [u8; 32],
    pub device_cert_hash: [u8; 32],
    pub issuer_device_id: Vec<u8>,
    pub issuer_seq: u64,
    pub revocation_id: [u8; 16],
    pub reason_code: u8,
    pub created_at_ms: u64,
    pub signature: [u8; 64],
}

fn require_id(label: &str, raw: &[u8]) -> Result<(), String> {
    if raw.len() < ID_MIN || raw.len() > ID_MAX {
        return Err(format!("{label} length must be 1..64"));
    }
    Ok(())
}

fn require_address(addr: &str) -> Result<(), String> {
    if addr.len() != ADDRESS_LEN {
        return Err("identity_address must be 44 chars".into());
    }
    if decode_address(addr).is_none() {
        return Err("identity_address invalid".into());
    }
    if addr != addr.to_lowercase() || !addr.starts_with("rvn1") {
        return Err("identity_address not canonical".into());
    }
    Ok(())
}

impl DeviceRevocationV1 {
    pub fn signing_bytes(&self) -> Result<Vec<u8>, String> {
        require_address(&self.identity_address)?;
        require_id("device_id", &self.device_id)?;
        require_id("issuer_device_id", &self.issuer_device_id)?;
        if self.revocation_id == [0u8; 16] {
            return Err("revocation_id all-zero".into());
        }
        let mut out = Vec::new();
        out.extend_from_slice(SIGNING_DOMAIN);
        out.push(VERSION);
        out.push(SUITE);
        out.extend_from_slice(self.identity_address.as_bytes());
        out.extend(lp(&self.device_id)?);
        out.extend_from_slice(&self.device_ed_pub);
        out.extend_from_slice(&self.device_x_pub);
        out.extend_from_slice(&self.device_cert_hash);
        out.extend(lp(&self.issuer_device_id)?);
        out.extend_from_slice(&u64_be(self.issuer_seq));
        out.extend_from_slice(&self.revocation_id);
        out.push(self.reason_code);
        out.extend_from_slice(&u64_be(self.created_at_ms));
        Ok(out)
    }

    pub fn encode(&self) -> Result<Vec<u8>, String> {
        let sb = self.signing_bytes()?;
        let mut out = Vec::with_capacity(8 + sb.len() - SIGNING_DOMAIN.len() + 64);
        out.extend_from_slice(MAGIC);
        out.extend_from_slice(&sb[SIGNING_DOMAIN.len()..]);
        out.extend_from_slice(&self.signature);
        Ok(out)
    }

    pub fn decode(wire: &[u8]) -> Result<Self, String> {
        if wire.len() < 8 + 2 + ADDRESS_LEN + 2 + ID_MIN + 32 + 32 + 32 + 2 + ID_MIN + 8 + 16 + 1 + 8 + 64
        {
            return Err("wire too short".into());
        }
        if &wire[..8] != MAGIC {
            return Err("bad magic".into());
        }
        if wire[8] != VERSION || wire[9] != SUITE {
            return Err("bad version/suite".into());
        }
        let mut off = 10;
        let addr = std::str::from_utf8(&wire[off..off + ADDRESS_LEN])
            .map_err(|_| "identity_address not utf8")?;
        let identity_address = addr.to_string();
        require_address(&identity_address)?;
        off += ADDRESS_LEN;
        let (device_id, noff) = read_lp(wire, off)?;
        off = noff;
        let mut device_ed_pub = [0u8; 32];
        device_ed_pub.copy_from_slice(&wire[off..off + 32]);
        off += 32;
        let mut device_x_pub = [0u8; 32];
        device_x_pub.copy_from_slice(&wire[off..off + 32]);
        off += 32;
        let mut device_cert_hash = [0u8; 32];
        device_cert_hash.copy_from_slice(&wire[off..off + 32]);
        off += 32;
        let (issuer_device_id, noff) = read_lp(wire, off)?;
        off = noff;
        let issuer_seq = u64::from_be_bytes(wire[off..off + 8].try_into().unwrap());
        off += 8;
        let mut revocation_id = [0u8; 16];
        revocation_id.copy_from_slice(&wire[off..off + 16]);
        off += 16;
        if revocation_id == [0u8; 16] {
            return Err("revocation_id all-zero".into());
        }
        let reason_code = wire[off];
        off += 1;
        let created_at_ms = u64::from_be_bytes(wire[off..off + 8].try_into().unwrap());
        off += 8;
        let mut signature = [0u8; 64];
        signature.copy_from_slice(&wire[off..off + 64]);
        off += 64;
        if off != wire.len() {
            return Err("trailing bytes".into());
        }
        Ok(Self {
            identity_address,
            device_id,
            device_ed_pub,
            device_x_pub,
            device_cert_hash,
            issuer_device_id,
            issuer_seq,
            revocation_id,
            reason_code,
            created_at_ms,
            signature,
        })
    }

    pub fn sign(mut self, id: &Identity) -> Result<Self, String> {
        self.identity_address = id.address();
        let sb = self.signing_bytes()?;
        self.signature = id.sign(&sb);
        Ok(self)
    }

    pub fn verify(&self, identity_ed_pub: &[u8; 32]) -> Result<(), String> {
        if encode_address(identity_ed_pub) != self.identity_address {
            return Err("ADDR_MISMATCH".into());
        }
        let sb = self.signing_bytes()?;
        if !Identity::verify(identity_ed_pub, &sb, &self.signature) {
            return Err("BAD_SIG".into());
        }
        Ok(())
    }
}

fn read_lp(buf: &[u8], off: usize) -> Result<(Vec<u8>, usize), String> {
    if off + 2 > buf.len() {
        return Err("truncated lp".into());
    }
    let n = u16::from_be_bytes([buf[off], buf[off + 1]]) as usize;
    if !(ID_MIN..=ID_MAX).contains(&n) {
        return Err("lp length out of 1..64".into());
    }
    let end = off + 2 + n;
    if end > buf.len() {
        return Err("truncated lp body".into());
    }
    Ok((buf[off + 2..end].to_vec(), end))
}

pub fn claim_digest(wire: &[u8]) -> [u8; 32] {
    let mut h = Sha256::new();
    h.update(wire);
    h.finalize().into()
}

#[derive(Debug, Clone)]
pub struct StoreClaim {
    pub exact_record_bytes: Vec<u8>,
}

#[derive(Debug, Clone)]
pub struct ExhaustedMarker {
    pub identity_address: String,
    pub claim_digest: [u8; 32],
    pub exact_record_bytes: Vec<u8>,
}

#[derive(Debug, Clone)]
pub struct CorruptMarker {
    pub scope: String,
    pub reason_code: u8,
}

pub fn canonical_store_snapshot(
    generation: u64,
    claims: &[StoreClaim],
    exhausted: &[ExhaustedMarker],
    corrupt: &[CorruptMarker],
) -> Result<Vec<u8>, String> {
    let mut claims_sorted = claims.to_vec();
    claims_sorted.sort_by_key(|a| claim_digest(&a.exact_record_bytes));
    let mut exh = exhausted.to_vec();
    exh.sort_by(|a, b| {
        (
            a.identity_address.as_bytes(),
            a.claim_digest.as_slice(),
        )
            .cmp(&(b.identity_address.as_bytes(), b.claim_digest.as_slice()))
    });
    let mut cor = corrupt.to_vec();
    cor.sort_by_key(|a| a.scope.clone());

    let mut out = Vec::new();
    out.extend_from_slice(STORE_SNAPSHOT_DOMAIN);
    out.push(0);
    out.extend_from_slice(&u64_be(generation));
    out.extend_from_slice(&(claims_sorted.len() as u32).to_be_bytes());
    for c in &claims_sorted {
        let d = claim_digest(&c.exact_record_bytes);
        out.extend_from_slice(&d);
        out.extend_from_slice(&(c.exact_record_bytes.len() as u32).to_be_bytes());
        out.extend_from_slice(&c.exact_record_bytes);
    }
    out.extend_from_slice(&(exh.len() as u32).to_be_bytes());
    for e in &exh {
        if claim_digest(&e.exact_record_bytes) != e.claim_digest {
            return Err("exhausted digest mismatch".into());
        }
        out.extend(lp(e.identity_address.as_bytes())?);
        out.extend_from_slice(&e.claim_digest);
        out.extend_from_slice(&(e.exact_record_bytes.len() as u32).to_be_bytes());
        out.extend_from_slice(&e.exact_record_bytes);
    }
    out.extend_from_slice(&(cor.len() as u32).to_be_bytes());
    for c in &cor {
        out.extend(lp(c.scope.as_bytes())?);
        out.push(c.reason_code);
    }
    Ok(out)
}

pub fn revocation_store_hash(
    generation: u64,
    claims: &[StoreClaim],
    exhausted: &[ExhaustedMarker],
    corrupt: &[CorruptMarker],
) -> Result<[u8; 32], String> {
    let snap = canonical_store_snapshot(generation, claims, exhausted, corrupt)?;
    let mut h = Sha256::new();
    h.update(&snap);
    Ok(h.finalize().into())
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::Value;
    use std::path::PathBuf;

    fn vectors_root() -> PathBuf {
        let mut p = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
        p.pop();
        p.pop();
        p.pop();
        p.join("shared-vectors").join("rvn1")
    }

    fn load(rel: &str) -> Value {
        let raw = std::fs::read_to_string(vectors_root().join(rel)).unwrap();
        serde_json::from_str(&raw).unwrap()
    }

    fn hex(s: &str) -> Vec<u8> {
        hex::decode(s).unwrap()
    }

    #[test]
    fn valid_001_roundtrip() {
        let v = load("device_revocation/valid_001.json");
        let wire = hex(v["expected"]["wire_hex"].as_str().unwrap());
        let rec = DeviceRevocationV1::decode(&wire).unwrap();
        assert_eq!(rec.encode().unwrap(), wire);
        let mut pubk = [0u8; 32];
        pubk.copy_from_slice(&hex(v["inputs"]["identity_ed_pub_hex"].as_str().unwrap()));
        rec.verify(&pubk).unwrap();
        assert_eq!(
            hex::encode(claim_digest(&wire)),
            v["expected"]["claim_digest_hex"].as_str().unwrap()
        );
    }

    #[test]
    fn wrong_signer_rejects() {
        let v = load("negative/device_revocation_wrong_signer.json");
        let wire = hex(v["inputs"]["wire_hex"].as_str().unwrap());
        let rec = DeviceRevocationV1::decode(&wire).unwrap();
        let mut pubk = [0u8; 32];
        pubk.copy_from_slice(&hex(
            v["inputs"]["claimed_identity_ed_pub_hex"].as_str().unwrap(),
        ));
        assert!(rec.verify(&pubk).is_err());
    }

    #[test]
    fn store_hash_parity() {
        for name in [
            "device_revocation/store_hash_001.json",
            "device_revocation/store_hash_exhausted_001.json",
        ] {
            let v = load(name);
            let claims: Vec<StoreClaim> = v["inputs"]["claims_wire_hex"]
                .as_array()
                .unwrap()
                .iter()
                .map(|h| StoreClaim {
                    exact_record_bytes: hex(h.as_str().unwrap()),
                })
                .collect();
            let exhausted: Vec<ExhaustedMarker> = v["inputs"]["exhausted"]
                .as_array()
                .unwrap()
                .iter()
                .map(|e| {
                    let mut d = [0u8; 32];
                    d.copy_from_slice(&hex(e["claim_digest_hex"].as_str().unwrap()));
                    ExhaustedMarker {
                        identity_address: e["identity_address"].as_str().unwrap().to_string(),
                        claim_digest: d,
                        exact_record_bytes: hex(e["exact_record_bytes_hex"].as_str().unwrap()),
                    }
                })
                .collect();
            let gen = v["inputs"]["generation"].as_u64().unwrap();
            let h = revocation_store_hash(gen, &claims, &exhausted, &[]).unwrap();
            assert_eq!(
                hex::encode(h),
                v["expected"]["revocation_store_hash_hex"].as_str().unwrap()
            );
            let snap = canonical_store_snapshot(gen, &claims, &exhausted, &[]).unwrap();
            assert_eq!(
                hex::encode(snap),
                v["expected"]["canonical_snapshot_hex"].as_str().unwrap()
            );
        }
    }
}
